postprocessing/DownloadRunnable.java

```diff
@@ -0,0 +1,184 @@
+
+        try {
+            f = mMission.storage.getStream();
+        } catch (IOException e) {
+            mMission.notifyError(e);// this never should happen
+            return;
+        }
+
+        while (mMission.running && mMission.errCode == DownloadMission.ERROR_NOTHING) {
+            if (!retry) {
+                block = mMission.acquireBlock();
+            }
+
+            if (block == null) {
+                if (DEBUG) Log.d(TAG, mId + ":no more blocks left, exiting");
+                break;
+            }
+
+            if (DEBUG) {
+                if (retry)
+                    Log.d(TAG, mId + ":retry block at position=" + block.position + " from the start");
+                else
+                    Log.d(TAG, mId + ":acquired block at position=" + block.position + " done=" + block.done);
+            }
+
+            long start = (long)block.position * DownloadMission.BLOCK_SIZE;
+            long end = start + DownloadMission.BLOCK_SIZE - 1;
+
+            start += block.done;
+
+            if (end >= mMission.length) {
+                end = mMission.length - 1;
+            }
+
+            try {
+                mConn = mMission.openConnection(false, start, end);
+                mMission.establishConnection(mId, mConn);
+
+                // check if the download can be resumed
+                if (mConn.getResponseCode() == 416) {
+                    if (block.done > 0) {
+                        // try again from the start (of the block)
+                        mMission.notifyProgress(-block.done);
+                        block.done = 0;
+                        retry = true;
+                        mConn.disconnect();
+                        continue;
+                    }
+
+                    throw new DownloadMission.HttpError(416);
+                }
+
+                retry = false;
+
+                // The server may be ignoring the range request
+                if (mConn.getResponseCode() != 206) {
+                    if (DEBUG) {
+                        Log.e(TAG, mId + ":Unsupported " + mConn.getResponseCode());
+                    }
+                    mMission.notifyError(new DownloadMission.HttpError(mConn.getResponseCode()));
+                    break;
+                }
+
+                f.seek(mMission.offsets[mMission.current] + start);
+
+                try (InputStream is = mConn.getInputStream()) {
+                    byte[] buf = new byte[DownloadMission.BUFFER_SIZE];
+                    int len;
+
+                    // use always start <= end
+                    // fixes a deadlock because in some videos, youtube is sending one byte alone
+                    while (start <= end && mMission.running && (len = is.read(buf, 0, buf.length)) != -1) {
+                        f.write(buf, 0, len);
+                        start += len;
+                        block.done += len;
+                        mMission.notifyProgress(len);
+                    }
```

Comment:
<div>


<div id="suggestion">
<div id="issue"><b>Consider extracting download loop logic</b></div>
<div id="fix">

Consider extracting the download loop into a separate method to improve readability and maintainability. The current implementation has complex nested logic within the `run()` method.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -119,6 +119,10 @@
     private void downloadBlock(Block block, long start, long end, SharpStream f) throws IOException {
         try (InputStream is = mConn.getInputStream()) {
             byte[] buf = new byte[DownloadMission.BUFFER_SIZE];
             int len;
             while (start <= end && mMission.running && (len = is.read(buf, 0, buf.length)) != -1) {
                 f.write(buf, 0, len);
                 start += len;
                 block.done += len;
                 mMission.notifyProgress(len);
             }
         }
     }
```
</div>
</details>
</div>


<small><i>Code Review Run <a href=https://github.com/OBrutus/youtube-downloader/pull/7#issuecomment-2463682284>#c539b9</a></i></small>
</div>

---
Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged